### PR TITLE
Monitor that 2xx pages don't regress to 404s

### PR DESCRIPTION
**Pull request by bishboria**
We have recently experienced two incidents (2015-07-13: Collections & Mainstream browse and 2015-06-26: Whitehall 404 errors) where sections of GOV.UK started returning HTTP 404 responses for content which was previously valid. No automated monitoring picked up these incidents - we'd like to have automated monitoring if this happens in future.

See https://trello.com/c/hbGMQfZL/301-monitor-that-2xx-pages-don-t-regress-to-404s

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,5 +1,22 @@
...
+
+`./process_404s.sh /src/log/directory /path/to/masterlist.csv`
+* streams data from the latest uncompressed cdn-log and compares it against the known good pages on gov.uk
+
+`./owning_app.sh "/base/path"`
+* outputs the name of the application that 'owns' the html content at the specified basepath on gov.uk.
+
+`./accumulate.sh /src/csv/directory /path/to/masterlist.csv`
+* takes urls that are known to be good and adds them to the masterlist
+
```
Comment (tijmenb) [inline README.md]: Could you add an example to run with the sample logs?

File: README.md [hunk 2]
```diff
@@ -1,5 +1,22 @@
-# Monitor 404s
+# govuk-cdn-logs-monitor
```
Comment (tijmenb) [inline README.md]: Might be useful to keep this in line with the styleguide.

https://github.com/alphagov/styleguides/blob/master/use-of-READMEs.md

File: lib/accumulate_into_master.rb [hunk 3]
```diff
@@ -0,0 +1,39 @@
+require_relative 'config'
+
+_200s_directory = ARGV[0]
```
Comment (tijmenb) [inline lib/accumulate_into_master.rb]: Underscores are mostly used to signify unused variables - this is enforced by rubocop (don't know if it's in our own styleguide) https://github.com/bbatsov/rubocop/blob/master/config/enabled.yml#L1011

File: lib/accumulate_into_master.rb [hunk 4]
```diff
@@ -0,0 +1,39 @@
+require_relative 'config'
```
Comment (tijmenb) [inline lib/accumulate_into_master.rb]: Maybe we can explain what the script does here. Am I correct in thinking that this takes the server logs for the last week and produces a master-list of URLs that should never return 200?

File: lib/alert_if_404_url_present.rb [hunk 5]
```diff
@@ -0,0 +1,24 @@
+require_relative 'config'
+
+masterlist = ARGV[0]
+
+known_good = []
+CSV.foreach(masterlist, 'r') do |row| # assume each line sorted by basepath
+  known_good << row[0]
+end
+
+$stdin.each_line do |line|
```
Comment (tijmenb) [inline lib/alert_if_404_url_present.rb]: A comment with an example value for `line` here would make it clear what we can expect things like `fragment[-2]` to be.

File: lib/alert_if_404_url_present.rb [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+  begin
+    fragment = line.split
+  rescue ArgumentError # weird characters in url
+    next
+  end
+  next if fragment[-1] != "404"
+
+  _404_path = fragment[-2]
+  if known_good.any? {|x| x == _404_path}
```
Comment (tijmenb) [inline lib/alert_if_404_url_present.rb]: Would `known_good.include?(404_path)` be clearer?

File: lib/process_200s_from_cdn_log.rb [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+  status = fragment[-1]
+  basepath = fragment[-2]
+  if status[0] == "2"
+    data[basepath] += 1
+  end
+end
+
+CSV.open(filename,"w") do |csv|
+  data.each do |basepath, count|
+    if basepath.include?("/y/") && count < 100
```
Comment (tijmenb) [inline lib/process_200s_from_cdn_log.rb]: Why the y?

**Comment by KushalP** [comment]
I think I have a little more context on this work than other people do as a result of helping you understand our CDN logs.

I know this code "works", but given that it's going to be used to monitor (and then alert on) our HTTP responses I'd prefer it to have some tests. That way other people (or our future selves) have some context when/if the log format or `awk` indexing changes.

**Comment by bishboria** [comment]
@KushalP thanks for the `add tests` reminder!

**Pull request merged**